core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp

```diff
@@ -539,17 +539,9 @@ class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
         m_vector_size(arg_policy.impl_vector_length()) {
     auto internal_space_instance =
         m_policy.space().impl_internal_space_instance();
-    cudaFuncAttributes attr =
-        CudaParallelLaunch<ParallelFor, LaunchBounds>::get_cuda_func_attributes(
-            internal_space_instance->m_cudaDev);
-    m_team_size =
-        m_team_size >= 0
-            ? m_team_size
-            : Kokkos::Impl::cuda_get_opt_block_size<FunctorType, LaunchBounds>(
-                  internal_space_instance, attr, m_functor, m_vector_size,
-                  m_policy.team_scratch_size(0),
-                  m_policy.thread_scratch_size(0)) /
-                  m_vector_size;
+    m_team_size = m_team_size >= 0 ? m_team_size
+                                   : arg_policy.team_size_recommended(
```

Comment:
The changes looks good overall ! (I'm always happier with code reuse :+1:)

However, I wonder about one thing that this PR will change.

I think https://github.com/kokkos/kokkos/blob/9fff1e066a7c3f2759d251e327fc08963f4f8c2b/core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp#L144-L147 will set the team and thread scratch sizes regardless of whether it is a parallel for or a parallel reduce.

But I guess that for a parallel for, there is no need for the tiny additional scratch size.

Reply:
A possible fix would be to modify the recommended team size function along these lines:
```c++
template <class FunctorType, class ParallelTagType>
  int team_size_recommended(const FunctorType& f, const ParallelTagType&) const {
    using closure_type =
        Impl::ParallelFor<FunctorType, TeamPolicy<Properties...>>;
    cudaFuncAttributes attr =
        CudaParallelLaunch<closure_type, typename traits::launch_bounds>::
            get_cuda_func_attributes(space().cuda_device());
    const int block_size =
        Kokkos::Impl::cuda_get_opt_block_size<FunctorType,
                                              typename traits::launch_bounds>(
            space().impl_internal_space_instance(), attr, f,
            (size_t)impl_vector_length(),
            (size_t)team_scratch_size(0) + is it reduce ? 2 * sizeof(double) : 0,
            (size_t)thread_scratch_size(0) + is it reduce ? sizeof(double) : 0);
    return block_size / impl_vector_length();
  }
```

Reply:
What Daniel said.